openmoc/materialize.py

```diff
@@ -355,3 +360,328 @@ def load_openmc_mgxs_lib(mgxs_lib, geometry=None):
+
+            inner_sph = np.ones((num_domains, num_groups))
+
+            for j in range(max_fix_src_iters):
+
+                # Run fixed source calculation with suppressed output
+                if throttle_output and log_level <= openmoc.WARNING:
+                    openmoc.set_log_level('WARNING')
+
+                # Fixed source calculation
+                solver.computeFlux()
+
+                # Restore log output level
+                if throttle_output and log_level <= openmoc.NORMAL: 
+                    openmoc.set_log_level('NORMAL')
+
+                # Extract the FSR scalar fluxes
+                fsr_fluxes = get_scalar_fluxes(solver)
+
+                # Compute the domain-averaged flux in each energy group
+                for k, domain in enumerate(mgxs_lib.domains):
+                    domain_fluxes = fsr_fluxes[fsrs_to_domains == domain.id, :]
+                    openmoc_fluxes[k, :] = np.mean(domain_fluxes, axis=0)
+
+                    # Store index for the domain in the array of SPH factors
+                    if domain.id == domain_id:
+                        sph_index = k
+
+                # Compute SPH factors
+                old_inner_sph = np.copy(inner_sph)
+                inner_sph = openmc_fluxes / openmoc_fluxes
+                inner_sph = np.nan_to_num(inner_sph)
+                inner_sph[inner_sph == 0.0] = 1.0
+
+                # Compute SPH factor residuals
+                inner_res = np.abs((inner_sph - old_inner_sph) / old_inner_sph)
+                inner_res = np.nan_to_num(inner_res)
+
+                # Report maximum SPH factor residual
+                py_printf('NORMAL', 'SPH inner iteration %d:\tres = '
+                                    '%1.3e', j, inner_res.max())
+
+                # Update multi-group cross sections with SPH factors
+                sph_mgxs_lib = _apply_sph_factors(mgxs_lib, inner_sph[sph_index],
+                                                  sph_domains[domain_id])
+
+                # Load the new MGXS library data into the OpenMOC geometry
+                load_openmc_mgxs_lib(sph_mgxs_lib, geometry)
+
+                # Check max SPH factor residual for this domain for convergence
+                if inner_res[sph_index].max() < sph_tol:
+                    break
+
+            mgxs_lib = sph_mgxs_lib
+
+        # Check max SPH factor residuals for domains of interest for convergence
+        if inner_res[sph_indices].max() < sph_tol:
+            break
+
+    # Warn user if SPH factors did not converge
+    else:
+        py_printf('WARNING', 'SPH factors did not converge')
+
+    # Compute the final SPH factors for all domains and energy groups
+    final_sph = openmc_fluxes / openmoc_fluxes
+    final_sph = np.nan_to_num(final_sph)
+    final_sph[final_sph == 0.0] = 1.0
+
+    # Restore verbosity log level to original status
+    openmoc.set_log_level(log_level)
+
+    return final_sph, mgxs_lib
```

Comment:
With some nifty Pythonic tricks (or perhaps not so much), I've update this PR in f235f4f6f2215d5328243ae9c832ad8b9a7c3e9a to simultaneously update the MGXS for *all* fissionable domains in each inner iteration when using Hebert's original SPH scheme. I changed the parameter `domains` to `mode` which can take the values `"fissionable"` (default, Hebert's scheme) or `"all"` (kord's scheme). Take a look at this @PrezNattyGibbs and let me know what you think.